### PR TITLE
Cleaning Cppcheck warnings and refactor tests

### PR DESCRIPTION
**Pull request by scemama**
# Pointers
instead of `const char my_string[256] = "hello"`, we should use `const char* my_string = "hello"`

# index_p 

I created a const and a non_const variable. If a new array is allocated, we use index_p and free the non_const variable. If we can use the initial variable, index_p is NULL and we don't use the non_const variable.

# String overflow

If a struct has a fixed-size string + some data, making an overflow in the string will corrupt data of the struct. So it is recommended to put strings at the end of structs, so that important data are less likely to be overwritten by s string overflow.

### CODE REVIEW

**Comment by scemama** [comment]
# Pointers
instead of `const char my_string[256] = "hello"`, we should use `const char* my_string = "hello"`

# index_p 

I created a const and a non_const variable. If a new array is allocated, we use index_p and free the non_const variable. If we can use the initial variable, index_p is NULL and we don't use the non_const variable.

# String overflow

If a struct has a fixed-size string + some data, making an overflow in the string will corrupt data of the struct. So it is recommended to put strings at the end of structs, so that important data are less likely to be overwritten by s string overflow.

**Review by q-posev — COMMENTED** [review]
The string changes are very useful, thanks!

File: src/templates_hdf5/templator_hdf5.org [hunk 1]
```diff
@@ -1465,24 +1468,22 @@ trexio_hdf5_open_read_dset_sparse (const hid_t group_id,
       uint16_t* index = CALLOC(size_ranked, uint16_t);
       if (index == NULL) return TREXIO_ALLOCATION_FAILED;
       index_p = index;
-    } else {
-      index_p = data_sparse;
```
Comment (q-posev) [inline src/templates_hdf5/templator_hdf5.org:1469]: Why did you remove this branch?

File: src/templates_hdf5/templator_hdf5.org [hunk 2]
```diff
@@ -497,7 +497,8 @@ trexio_hdf5_write_$group_dset$ (trexio_t* const file,
   trexio_hdf5_t* f = (trexio_hdf5_t*) file;
 
   hid_t index_dtype;
-  void* index_p = NULL;
+  const void* index_p = NULL;
```
Comment (q-posev) [inline src/templates_hdf5/templator_hdf5.org:500]: Could you clarify a bit why we need 2 void pointers here instead of one as it was implemented before?

Comment (scemama) [reply src/templates_hdf5/templator_hdf5.org:500]: `index_sparse` is a `const int32_t *`. 
What we did before was to create a new array for smaller types, and use the same pointer if we are in 32 bits.
The problem is that if we create a new pointer and make a new malloc, we need to lose the `const` attribute because we need to free the memory.
So we create a const pointer and a non-const pointer. The const pointer takes the value whether of `index_sparse` or the new allocated pointer, and we can happily pass this `const` pointer to the `write` routines that expect a const pointer. But to free the memory, we use the non-const pointer.

File: src/templates_hdf5/templator_hdf5.org (hunk 1 above)
Comment (scemama) [reply src/templates_hdf5/templator_hdf5.org:1469]: Now `index_p` is initialized to `NULL` and used only if we are not using 32 bits. So below we check if `index_p` is `NULL`. If it is, we use the `index_sparse` pointer. It avoids the problem of having `index_p` that is sometimes  `const` and sometimes not. Now, `index_p` is never `const`, and we use `index_sparse` when needed.

Comment (q-posev) [reply src/templates_hdf5/templator_hdf5.org:1469]: I think you are wrong. This modification disables read of sparse indices in 32-bits precision. The unit tests do not catch this probably because the test number of MOs is small and uses lower-bit indices.

This function is generic, that's why the input data pointer is void*. It is called 2 times: first for indices and then for values. In general, i would prefer to avoid messing up with the logic of sparse data I/O which is already quite complex.

File: src/templates_hdf5/templator_hdf5.org (hunk 2 above)
Comment (q-posev) [reply src/templates_hdf5/templator_hdf5.org:500]: Thanks, now i see. Have you tested this for I/O of indices with large enough number of MOs (so that 32bit precision is actually used)? The unit tests use dummy 1000 mo-num so the 32bit code path is not tested.

File: src/templates_hdf5/templator_hdf5.org (hunk 1 above)
Comment (scemama) [reply src/templates_hdf5/templator_hdf5.org:1469]: Below, we have:
```
  if (is_index == 1) {
    status = H5Dread(dset_id,
                     dtype,
                     memspace_id, fspace_id, H5P_DEFAULT,
                     index_p);
  } else {
    status = H5Dread(dset_id,
                     dtype,
                     memspace_id, fspace_id, H5P_DEFAULT,
                     data_sparse);
  }
```
so we use `data_sparse` if `is_index` is not 1. It seems OK to me.

Also, in my last push, I implemented tests for different numbers of MOs: 10, 20, 40, 100, 300.
They all seem to be OK.

File: src/templates_hdf5/templator_hdf5.org (hunk 2 above)
Comment (scemama) [reply src/templates_hdf5/templator_hdf5.org:500]: Tested in the last commit

**Comment by scemama** [comment]
@q-posev ping!

**Comment by q-posev** [comment]
> @q-posev ping!

@scemama I responded a few days ago in [this thread](https://github.com/TREX-CoE/trexio/pull/115#discussion_r1148525424). I still believe that changes in this PR break the I/O of indices in 32-bit, which is why I am not ready to merge it.

**Review by q-posev — COMMENTED** [review]
@scemama Sorry, the review was marked as `pending` for some reason by GitHub... You should be able to see it now

**Comment by q-posev** [comment]
@scemama ping 🙂

**Comment by q-posev** [comment]
@scemama thanks! I think the name of the PR can be changed to reflect that there was a major refactoring of the unit tests. I have one comment, let me know if you prefer to get it fixed in a separate PR or in this one:

We discussed testing the `io_dset_sparse` with different values of `mo_num` in order to test at least the 8/16-bit and 32-bit code paths.

**Comment by scemama** [comment]
> We discussed testing the io_dset_sparse with different values of mo_num in order to test at least the 8/16-bit and 32-bit code paths.

Done!

**Review by q-posev — APPROVED** [review]
